Source/WTF/wtf/Dominators.h

```diff
@@ -294,6 +314,129 @@ class Dominators {
     }
     
 private:
+    // This implements Cooper, Harvey, and Kennedy's iterative dominance algorithm as described in
+    // "A Simple, Fast Dominance Algorithm" (2001). Compared to Lengauer and Tarjan's method, which is
+    // O(n log n), the iterative method is O(N + E * D), where D is the size of the set of dominators
+    // for a particular node. This is worst-case quadratic, but likely better in practice for real code
+    // where the average number of dominators does not grow nearly as fast as the number of nodes.
+    // Moreover, this algorithm is much simpler, requiring very little auxiliary data and generally
+    // having substantially better constant factors. We prefer this algorithm for most graphs, the
+    // asymptotic complexity only becoming an issue for very large functions (10000s of blocks).
+    // https://www.clear.rice.edu/comp512/Lectures/Papers/TR06-33870-Dom.pdf
+
+    class IterativeDominance {
+        WTF_MAKE_FAST_ALLOCATED;
+        constexpr static uint16_t undefinedIdom = std::numeric_limits<uint16_t>::max();
+    public:
+        IterativeDominance(Graph& graph)
+            : m_graph(graph)
+        {
+            // We only use this for small-ish graphs. So, we exploit that to use
+            // smaller integers for idom information. We mostly use uint16_t for
+            // our analysis, but we exploit int16_t when computing reverse
+            // postorder. We expect Lengauer-Tarjan to beat us beyond a few ten
+            // thousand blocks anyway so this should be fine.
+            RELEASE_ASSERT(graph.numNodes() < std::numeric_limits<int16_t>::max());
+            m_idoms.fill(undefinedIdom, graph.numNodes());
+        }
+
+        void computeReversePostorder()
+        {
+            BitSet<std::numeric_limits<int16_t>::max()> visited;
+            visited.clearAll();
+
+            Vector<int16_t> workList;
+            int16_t rootIndex = m_graph.index(m_graph.root());
+            workList.append(rootIndex);
+            visited.set(rootIndex);
+
+            while (workList.size()) {
+                int16_t index = workList.takeLast();
+                if (index < 0) {
+                    // Negative indices mark nodes we're revisiting.
+                    m_order.append(~index);
+                    continue;
+                }
+                const auto& successors = m_graph.successors(m_graph.node(index));
+                if (!successors.size()) {
```

Comment:
`successors.isEmpty()`

Reply:
Actually I don't think that's available - `successors` is not a `Vector` but a `SuccessorList`, which is not assumed to have an `isEmpty()` method. So I think we have to check the `size()` here - it's also what the existing dominance algorithm does.